cip/CIP2016-06-14-Harmonize-comparability-orderability-equality-and-equivalence.adoc

```diff
@@ -0,0 +1,233 @@
+* Comparing any kind of value not covered above to an arbitrary other value, always yields `null`. This includes comparing a non-`null` value to a `null` value.
+
+=== Orderability
+
+We further propose that orderability be defined between any pair of values, where the result is always `true` or `false`; i.e. always defined.
+To accomplish this, there must be a pre-determined type order and each value must fall under exactly one type in this order. We propose to use the following ascending global sort order of types:
+
+* `NODE`
+* `RELATIONSHIP`
+* `PATH`
+* `MAP`
+* `LIST OF ANY?`
+* `STRING`
+* `BOOLEAN`
+* `NUMBER` (`NaN` is treated as the largest number in orderability only, i.e. it is put after positive infinity)
+* `VOID` (i.e. the type of `null`)
+
+Within the types, orderability defers to comparability with two exceptions:
+* Two `null` values are treated as the same value in accordance with the global sort order
+* Two `NaN` values are treated as the same value in accordance with the global sort order
+
+The accompanying descending global sort oder is the same order in reverse (i.e. it runs from `VOID` to `NODE`).
+
+Any additionally introduced non-canonical types introduced by an implementation should be inserted between `PATH` and `MAP` in the global sort order. As an exception, additionally introduced non-canonical number types may be integrated under the `NUMBER` type in the global sort order.
+
+=== Equality ===
+
+We propose that equality and comparability must be aligned, i.e. `l = r` if and only if `l <= r && l >= r`.
+
+To achieve this, we unify the difference between equality and equivalence, by making lists containing `null` values follow the same difference as the `null` values themselves.
+
+Concretely, we propose to redefine how equality works for lists in Cypher today.
+To determine if two lists `l1` and `l2` are equal, we propose two simple tests, like so
+
+* `l1` and `l2` must have the same size, i.e. inversely `size(l1) <> size(l2>) => l1 <> l2`
+* the pairwise elements of both `l1` and `l2` must be equal, i.e.
+----
+[a1, a2, ...] = [b1, b2, ...]
+<=>
+a1 = b1 && a2 = b2 && ...
+----
+
+For clarity, we also repeat the current equality semantics of maps here (at the time of writing this CIP). Two maps `m1` and `m2` are considered equal if
+
+* `m1` and `m2` must have the same keys.
+* For each key `k`,
+** either `m1.k = m2.k` is `true`,
+** or both `m1.k IS NULL` and `m2.k IS NULL`
+
+This is at odds with the common interpretation of `null` as standing for any possible value.
+
+It is aligned though with the most common use case for maps with `null` entries which is to update multiple properties at once, e.g. `SET n += { n: 12, remove_this_key: null }`. In this case, there is no need to differentiate between different `null` values as `null` merely serves as a marker for keys to be removed. Current equality semantics makes it easy to check if two maps correspond to the same property update in this scenario.
```

Comment:
"It is aligned though with the most common use case"
-->
"However, our interpretation of `null` is aligned with the most common use case"